# Clint Dozier BDI-475 Final Project
*Examining State of Illinois Employee Salaries Under The IL Central Management System (CMS), 2011-2021*

In [1]:
# IMPORT STATEMENTS
import pandas as pd
import numpy as np
import plotly
import plotly.express as px
import plotly.graph_objects as go

# plotly.io is a low-level interface for interacting with figures/
# plotly.io.templates lists available plotly templates
# https://plotly.com/python-api-reference/plotly.io.html
import plotly.io as pio

#pd.set_option('display.max_columns', 50)

In [2]:
# Read in the dataset

df = pd.read_csv('/content/drive/MyDrive/Spring 2021/BDI-475/il_employee_salaries_since_2011.csv')

# Delete redundant column

del df['_id']

df

Id  ...  Military Veteran
0       E66E4A4E767B69A7985744A673FD68AE59C9BCE3  ...                NO
1       B1C52532026C0EE32F21B60D4F76E9947CD47AAF  ...                NO
2       4F7CD61A17B87FB5062A02C79DFE795588295847  ...                NO
3       CC3EC019C668B48ED6C09BF1F18407787670A522  ...                NO
4       374B2A4CA9AC783EB415BA5E4B93908B1D0DBCC9  ...               YES
...                                          ...  ...               ...
183949  F62AFA20EF8110AB7F88507CF54EE77D7E11ABB3  ...                NO
183950  4E5E4401E79B4B38888B8DCC6502002281315088  ...                NO
183951  806D8D09A6BC4E662DBFD4DE06EC4902EE5827D0  ...                NO
183952  4D8FA26D4FA17E0C2604CD80D74FDC8B5F1D052C  ...                NO
183953  A18D469830FFAF1ADC8AAA318463BCD1AC3FC5BE  ...                NO

[183954 rows x 14 columns]

In [3]:
df.shape

(183954, 14)

This dataset contains 183,954 rows and 14 columns, and spans the course of 10 years, from 2011 to now.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183954 entries, 0 to 183953
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   Id                   183954 non-null  object
 1   Year                 183954 non-null  int64 
 2   Agency               183954 non-null  object
 3   Agency Division      183570 non-null  object
 4   Employee Name        183954 non-null  object
 5   Position Title       183838 non-null  object
 6   Period Pay Rate      183954 non-null  int64 
 7   YTD Gross            183954 non-null  int64 
 8   Work County          183954 non-null  object
 9   Rutan Exempt         183954 non-null  object
 10  Collective or Merit  183954 non-null  object
 11  4(d) Exempt          183954 non-null  object
 12  Employment Status    183950 non-null  object
 13  Military Veteran     183954 non-null  object
dtypes: int64(3), object(11)
memory usage: 19.6+ MB


In [5]:
# Group by YEAR and MEDIAN PERIOD PAY RATE

df_med_by_year = df.groupby('Year', as_index=False).agg({
    'Period Pay Rate': 'median'
}).rename(columns={
    'Period Pay Rate': 'Median Period Pay Rate'
})
df_med_by_year = df_med_by_year.sort_values(by='Median Period Pay Rate')

df_med_by_year

Year  Median Period Pay Rate
0   2011                    2000
1   2012                    2100
4   2015                    2200
5   2016                    2200
6   2017                    2200
7   2018                    2200
2   2013                    2300
3   2014                    2300
8   2019                    2300
9   2020                    2500
10  2021                    2600

In [6]:
# Distribution of Median Salaries by YEAR

fig1 = px.bar(df_med_by_year,
              x="Year",
              y="Median Period Pay Rate",
              template='plotly_dark',
              height = 800,
              title = 'Distribution of Median Salaries by Year'
              )

fig1

In [7]:
# Group by YEAR and AVG PERIOD PAY RATE

df_avg_by_year = df.groupby('Year', as_index=False).agg({
    'Period Pay Rate': 'mean'
}).rename(columns={
    'Period Pay Rate': 'Average Period Pay Rate'
})
df_avg_by_year = df_avg_by_year.sort_values(by='Average Period Pay Rate')

df_avg_by_year

Year  Average Period Pay Rate
0   2011              2049.547632
1   2012              2265.361680
5   2016              2346.570117
6   2017              2351.680450
4   2015              2360.071185
7   2018              2363.033096
2   2013              2394.744504
3   2014              2474.179871
8   2019              2591.938604
9   2020              2635.484492
10  2021              2712.894084

In [8]:
# Distribution of Avg Salaries by Year

fig2 = px.bar(df_avg_by_year,
              x="Year",
              y="Average Period Pay Rate",
              template='plotly_dark',
              height = 800,
              title = 'Distribution of Average Salaries by Year'
              )

fig2

Very interesting from both of these datasets. We can see that both median and average state of Illinois employee salaries remained entirely falt from 2015 through 2018. This timeframe coincides when Bruce Rauner, who was critical of state employee benefits that he deemed as overly generous, was Governor of Illinois.

In [9]:
# Group by AGENCY and MEDIAN PERIOD PAY RATE

df_agency_med = df.groupby('Agency', as_index=False).agg({
    'Period Pay Rate': 'median'
}).rename(columns={
    'Period Pay Rate': 'Median Period Pay Rate'
})
df_agency_med = df_agency_med.sort_values(by='Median Period Pay Rate')

df_agency_med

Agency  Median Period Pay Rate
58                    RACING BOARD                    1600
55           PRISONER REVIEW BOARD                    1600
31      IL COMMUNITY COLLEGE BOARD                    1800
51                MILITARY AFFAIRS                    2000
25    HEALTHCARE & FAMILY SERVICES                    2000
..                             ...                     ...
15  COUNCIL ON DEVELOPMENTAL DISAB                    4000
44           ILLINOIS POWER AGENCY                    4100
39      IL STATE TOLL HIGHWAY AUTH                    4200
8         CIVIL SERVICE COMMISSION                    4800
62        STATE BOARD OF ELECTIONS                    4900

[71 rows x 2 columns]

In [10]:
# Distribution of Median Salaries by Agency

fig3 = px.bar(df_agency_med,
              x="Agency",
              y="Median Period Pay Rate",
              template='plotly_dark',
              color='Median Period Pay Rate',
              color_continuous_scale = ['LightBlue', 'DodgerBlue'],
              height = 800,
              title = 'Distribution of Median Pay Rates by IL State Agency'
              )

fig3

In [11]:
# Group by COUNTY and MEDIAN PERIOD PAY RATE

df_county_med = df.groupby('Work County', as_index=False).agg({
    'Period Pay Rate': 'median'
}).rename(columns={
    'Period Pay Rate': 'Median Period Pay Rate'
})
df_county_med = df_county_med.sort_values(by='Median Period Pay Rate')

df_county_med

Work County  Median Period Pay Rate
43    KANKAKEE                    1800
30    HAMILTON                    1800
77      PUTNAM                    1850
75        POPE                    1900
0        ADAMS                    2000
..         ...                     ...
19     DE WITT                    2700
40  JO DAVIESS                    2800
96       WAYNE                    2800
3        BOONE                    3200
6      CALHOUN                    3200

[103 rows x 2 columns]

In [12]:
# Distribution of Median Salaries by County

fig4 = px.bar(df_county_med,
              x="Work County",
              y="Median Period Pay Rate",
              template='plotly_dark',
              color='Median Period Pay Rate',
              color_continuous_scale = ['LightBlue', 'DodgerBlue'],
              height = 800,
              title = 'Distribution of Median Pay Rates by IL County'
              )

fig4

In [13]:
# Group by COUNTY and NUMBER of 2020 EMPLOYEES

df_2020 = df[df['Year'] == 2020]

df_county_num = df_2020.groupby('Work County').size().reset_index(name='Number of Employees')
df_county_num = df_county_num.sort_values(by='Number of Employees')

df_county_num = df_county_num.rename(columns = {'Work County':'County'})

df_county_num

County  Number of Employees
32    HARDIN                    2
77    PUTNAM                    2
75      POPE                    2
64    MERCER                    2
3      BOONE                    3
..       ...                  ...
78  RANDOLPH                 1068
43  KANKAKEE                 1214
98      WILL                 1477
15      COOK                 5686
82  SANGAMON                 6116

[102 rows x 2 columns]

In [14]:
# Distribution of Employees by County (2021)

fig5 = px.bar(df_county_num,
              x="County",
              y="Number of Employees",
              template='plotly_dark',
              color='Number of Employees',
              color_continuous_scale = ['LightBlue', 'DodgerBlue'],
              height = 800,
              title = 'Distribution of Workers by County (2021)'
              )

fig5

In [15]:
# Read in dataset of 2020 population estimates for each Illinois county

df_pop_by_county = pd.read_csv('/content/drive/MyDrive/Spring 2021/BDI-475/Il_pop_by_county.csv')


#Capitalize the names of the counties

df_pop_by_county['County'] = df_pop_by_county['County'].str.upper()


df_pop_by_county.shape

(102, 2)

In [16]:
import math

In [17]:

# Merge the two datasets
df_pop_and_county = pd.merge(df_pop_by_county, df_county_num, on='County', how='outer')

# Add column of employees per 1000 residents
df_pop_and_county["Employees per Thousand Residents"] = ((df_pop_and_county["Number of Employees"] / df_pop_and_county["Population"]) * 1000)

# Sort this new column
df_pop_and_county = df_pop_and_county.sort_values(by='Employees per Thousand Residents')

# Drop NaN values (*Doesn't work*) HELP!
# df_pop_and_county = df_pop_and_county[df_pop_and_county['Employees per Thousand Residents'] != math.nan]

df_pop_and_county

County  ...  Employees per Thousand Residents
3           BOONE  ...                          0.048778
46        KENDALL  ...                          0.112031
65         MERCER  ...                          0.125810
55        MCHENRY  ...                          0.162292
12           CLAY  ...                          0.227756
..            ...  ...                               ...
23        EDWARDS  ...                               NaN
29       GALLATIN  ...                               NaN
102          CASS  ...                               NaN
103  OUT OF STATE  ...                               NaN
104       LASALLE  ...                               NaN

[105 rows x 4 columns]

In [18]:
# Distribution of Employees per Thousand Residents by County

fig6 = px.bar(df_pop_and_county,
              x="County",
              y="Employees per Thousand Residents",
              template='plotly_dark',
              color='Employees per Thousand Residents',
              color_continuous_scale = ['LightBlue', 'DodgerBlue'],
              height = 800,
              title = 'State of IL Employees per Thousand Residents by County'
              )

fig6

In [19]:
print('Restart and run all successful :)')

Restart and run all successful :)
